### PR TITLE
[lcn] Add action for lcn relay timer

### PR DESCRIPTION
**Pull request by toweosp**
This pull request adds an action for an lcn relay timer.

### CODE REVIEW

File: bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java [hunk 1]
```diff
@@ -777,4 +797,28 @@ private static int timeToRampValue(int timeMSec) {
...
     }
+
+    /**
+     * Converts duration in milliseconds to lcntimer value
+     * Source: https://www.symcon.de/forum/threads/38603-LCN-Relais-Kurzzeit-Timer-umrechnen
+     *
+     * @param ms time in milliseconds
+     * @return lcn timer value
+     */
+    private static Integer convertMsecToLCNTimer(double ms) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java:808]: Primitive types should be used where possible.
```suggestion
    private static int convertMsecToLCNTimer(double ms) {
```

File: bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java [hunk 2]
```diff
@@ -777,4 +797,28 @@ private static int timeToRampValue(int timeMSec) {
...
+        Integer lcntimer = -1;
+        if (ms >= 0 && ms <= 240960) {
+            double a = ms / 1000 / 0.03;
+            double b = (a / 32.0) + 1.0;
+            double c = Math.log(b) / Math.log(2);
+            double mod = Math.floor(c);
+            double faktor = 32 * (Math.pow(2, mod) - 1);
+            double offset = (a - faktor) / Math.pow(2, mod);
+            lcntimer = (new Double(offset + (mod * 32))).intValue();
```
Comment (fwolter) [inline bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java:817]: This should be sematically the same:
```suggestion
            lcntimer = (int) (offset + mod * 32);
```

File: bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java [hunk 3]
```diff
@@ -777,4 +797,28 @@ private static int timeToRampValue(int timeMSec) {
...
+        } else {
+            LOGGER.error("Timer not in [0,240960] ms");
```
Comment (fwolter) [inline bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java:819]: Bindings should only log to error if something severe happened, like the detection of a bug in the code. You could make this warn.

File: bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java [hunk 4]
```diff
@@ -777,4 +797,28 @@ private static int timeToRampValue(int timeMSec) {
         }
         return ret;
     }
+
+    /**
+     * Converts duration in milliseconds to lcntimer value
+     * Source: https://www.symcon.de/forum/threads/38603-LCN-Relais-Kurzzeit-Timer-umrechnen
```
Comment (fwolter) [inline bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java:803]: I honestly don't know how to handle this from the license perspective. I'll clarify this.

**Comment by timbms** [comment]
Relay timer is a command that was introduced in 2006. My LCN installation is older. Do you know whether it is dependent only on the LCH-HU or whether also the connected relay block would need to be replaced?

**Comment by fwolter** [comment]
> 
> 
> Relay timer is a command that was introduced in 2006. My LCN installation is older. Do you know whether it is dependent only on the LCH-HU or whether also the connected relay block would need to be replaced?

It's only a matter of the module's firmware.

File: bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java (hunk 3 above)
Comment (toweosp) [reply bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java]: @fwolter Thanks for your hints and suggestions. I have changed the log level to warn. Furthermore I corrected upper bound regarding duration in startRelayTimer to 240960 (was 240000).

Hopefully the formatting issues where fixed too.

**Comment by toweosp** [comment]
@fwolter could you give me a hint how to solve the travis build error with jdk11?

**Comment by kaikreuzer** [comment]
@toweosp You can ignore the Java11 build right now - this started happening for all builds recently, we will have to look into that.
Pretty sure it is not related to your PR.

File: bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java (hunk 4 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/common/PckGenerator.java:803]: I'd say it is ok. The code is not copied from there, but clearly self-written as it is a different programming language - so we don't have any copyright issues here.

**Pull request merged**